### ISSUE TITLE
InternalCompilerError: FixedPointType not implemented

### ISSUE THREAD

**Issue by pietroborrello** [ethereum/solidity#12972]
##### **Describe the bug**
The attached testcase crashes the solidity compiler solc with an InternalCompilerError: FixedPointType not implemented

##### **To Reproduce**
Built solc_ossfuzz using clang-10 according to [the oss-fuzz script](https://github.com/google/oss-fuzz/blob/master/projects/solidity/build.sh) with `CXXFLAGS='-O1 -fsanitize=address -fsanitize=array-bounds,bool,builtin,enum,float-divide-by-zero,function,integer-divide-by-zero,null,object-size,return,returns-nonnull-attribute,shift,signed-integer-overflow,unreachable,vla-bound,vptr'`

commit: b35cda5998ac7bc28a58f23480f2f82a440d6fa2

##### **Crash Output**

```
$ ./solc_ossfuzz id:000000,sig:06,src:012328,time:9919156,op:havoc,rep:2,trial:1
INFO: Seed: 2392507469
INFO: Loaded 1 modules   (537599 inline 8-bit counters): 537599 [0x53258a0, 0x53a8c9f), 
INFO: Loaded 1 PC tables (537599 PCs): 537599 [0x53a8ca0,0x5bdcc90), 
solc_ossfuzz: Running 1 inputs 1 time(s) each.
Running: id:000000,sig:06,src:012328,time:9919156,op:havoc,rep:2,trial:1
terminate called after throwing an instance of 'boost::wrapexcept<solidity::langutil::InternalCompilerError>'
  what():  FixedPointType not implemented.
==2678859== ERROR: libFuzzer: deadly signal
    #0 0x505691 in __sanitizer_print_stack_trace (solc_ossfuzz+0x505691)
    #1 0x55da28 in fuzzer::PrintStackTrace() (solc_ossfuzz+0x55da28)
    #2 0x542819 in fuzzer::Fuzzer::CrashCallback() (solc_ossfuzz+0x542819)
    #3 0x7ffff7c713bf  (/lib/x86_64-linux-gnu/libpthread.so.0+0x143bf)
    #4 0x7ffff7a8103a in __libc_signal_restore_set /build/glibc-sMfBJT/glibc-2.31/signal/../sysdeps/unix/sysv/linux/internal-signals.h:86:3
    #5 0x7ffff7a8103a in raise /build/glibc-sMfBJT/glibc-2.31/signal/../sysdeps/unix/sysv/linux/raise.c:48:3
    #6 0x7ffff7a60858 in abort /build/glibc-sMfBJT/glibc-2.31/stdlib/abort.c:79:7
    #7 0x7ffff7e6d910  (/lib/x86_64-linux-gnu/libstdc++.so.6+0x9e910)
    #8 0x7ffff7e7938b  (/lib/x86_64-linux-gnu/libstdc++.so.6+0xaa38b)
    #9 0x7ffff7e793f6 in std::terminate() (/lib/x86_64-linux-gnu/libstdc++.so.6+0xaa3f6)
    #10 0x7ffff7e796a8 in __cxa_throw (/lib/x86_64-linux-gnu/libstdc++.so.6+0xaa6a8)
    #11 0x6407ae in void boost::throw_exception<solidity::langutil::InternalCompilerError>(solidity::langutil::InternalCompilerError const&) (solc_ossfuzz+0x6407ae)
    #12 0xe4014e in solidity::frontend::TypeChecker::visit(solidity::frontend::InlineAssembly const&)::$_4::operator()(solidity::yul::Identifier const&, solidity::yul::IdentifierContext, bool) const (solc_ossfuzz+0xe4014e)
    #13 0x1a90806 in std::function<bool (solidity::yul::Identifier const&, solidity::yul::IdentifierContext, bool)>::operator()(solidity::yul::Identifier const&, solidity::yul::IdentifierContext, bool) const (solc_ossfuzz+0x1a90806)
    #14 0x1a8fd6e in solidity::yul::AsmAnalyzer::operator()(solidity::yul::Identifier const&) (solc_ossfuzz+0x1a8fd6e)
    #15 0x1aa76d0 in std::vector<solidity::yul::YulString, std::allocator<solidity::yul::YulString> > std::__invoke_impl<std::vector<solidity::yul::YulString, std::allocator<solidity::yul::YulString> >, solidity::yul::AsmAnalyzer&, solidity::yul::Identifier&>(std::__invoke_other, solidity::yul::AsmAnalyzer&, solidity::yul::Identifier&) (solc_ossfuzz+0x1aa76d0)
    #16 0x1aa7690 in std::__invoke_result<solidity::yul::AsmAnalyzer&, solidity::yul::Identifier&>::type std::__invoke<solidity::yul::AsmAnalyzer&, solidity::yul::Identifier&>(solidity::yul::AsmAnalyzer&, solidity::yul::Identifier&) (solc_ossfuzz+0x1aa7690)
    #17 0x1aa7652 in std::__detail::__variant::__gen_vtable_impl<true, std::__detail::__variant::_Multi_array<std::vector<solidity::yul::YulString, std::allocator<solidity::yul::YulString> > (*)(solidity::yul::AsmAnalyzer&, std::variant<solidity::yul::FunctionCall, solidity::yul::Identifier, solidity::yul::Literal>&)>, std::tuple<std::variant<solidity::yul::FunctionCall, solidity::yul::Identifier, solidity::yul::Literal>&>, std::integer_sequence<unsigned long, 1ul> >::__visit_invoke_impl(solidity::yul::AsmAnalyzer&, std::variant<solidity::yul::FunctionCall, solidity::yul::Identifier, solidity::yul::Literal>&) (solc_ossfuzz+0x1aa7652)
    #18 0x1aa7600 in std::__detail::__variant::__gen_vtable_impl<true, std::__detail::__variant::_Multi_array<std::vector<solidity::yul::YulString, std::allocator<solidity::yul::YulString> > (*)(solidity::yul::AsmAnalyzer&, std::variant<solidity::yul::FunctionCall, solidity::yul::Identifier, solidity::yul::Literal>&)>, std::tuple<std::variant<solidity::yul::FunctionCall, solidity::yul::Identifier, solidity::yul::Literal>&>, std::integer_sequence<unsigned long, 1ul> >::__do_visit_invoke(solidity::yul::AsmAnalyzer&, std::variant<solidity::yul::FunctionCall, solidity::yul::Identifier, solidity::yul::Literal>&) (solc_ossfuzz+0x1aa7600)
    #19 0x1aa7470 in std::__detail::__variant::__gen_vtable_impl<true, std::__detail::__variant::_Multi_array<std::vector<solidity::yul::YulString, std::allocator<solidity::yul::YulString> > (*)(solidity::yul::AsmAnalyzer&, std::variant<solidity::yul::FunctionCall, solidity::yul::Identifier, solidity::yul::Literal>&)>, std::tuple<std::variant<solidity::yul::FunctionCall, solidity::yul::Identifier, solidity::yul::Literal>&>, std::integer_sequence<unsigned long, 1ul> >::__visit_invoke(solidity::yul::AsmAnalyzer&, std::variant<solidity::yul::FunctionCall, solidity::yul::Identifier, solidity::yul::Literal>&) (solc_ossfuzz+0x1aa7470)
    #20 0x1aa73ac in decltype(auto) std::__do_visit<false, true, solidity::yul::AsmAnalyzer&, std::variant<solidity::yul::FunctionCall, solidity::yul::Identifier, solidity::yul::Literal>&>(solidity::yul::AsmAnalyzer&, std::variant<solidity::yul::FunctionCall, solidity::yul::Identifier, solidity::yul::Literal>&) (solc_ossfuzz+0x1aa73ac)
    #21 0x1a933a5 in decltype(auto) std::visit<solidity::yul::AsmAnalyzer&, std::variant<solidity::yul::FunctionCall, solidity::yul::Identifier, solidity::yul::Literal>&>(solidity::yul::AsmAnalyzer&, std::variant<solidity::yul::FunctionCall, solidity::yul::Identifier, solidity::yul::Literal>&) (solc_ossfuzz+0x1a933a5)
    #22 0x1a957b7 in solidity::yul::AsmAnalyzer::operator()(solidity::yul::VariableDeclaration const&) (solc_ossfuzz+0x1a957b7)
    #23 0x1a8bfed in solidity::yul::AsmAnalyzer::operator()(solidity::yul::Block const&) (solc_ossfuzz+0x1a8bfed)
    #24 0x1a8b7bf in solidity::yul::AsmAnalyzer::analyze(solidity::yul::Block const&) (solc_ossfuzz+0x1a8b7bf)
    #25 0xd91584 in solidity::frontend::TypeChecker::visit(solidity::frontend::InlineAssembly const&) (solc_ossfuzz+0xd91584)
    #26 0xe7a73e in solidity::frontend::InlineAssembly::accept(solidity::frontend::ASTConstVisitor&) const (solc_ossfuzz+0xe7a73e)
    #27 0xe7b79a in void solidity::frontend::ASTNode::listAccept<std::shared_ptr<solidity::frontend::Statement> >(std::vector<std::shared_ptr<solidity::frontend::Statement>, std::allocator<std::shared_ptr<solidity::frontend::Statement> > > const&, solidity::frontend::ASTConstVisitor&) (solc_ossfuzz+0xe7b79a)
    #28 0xe7b206 in solidity::frontend::Block::accept(solidity::frontend::ASTConstVisitor&) const (solc_ossfuzz+0xe7b206)
    #29 0xd6f07f in solidity::frontend::TypeChecker::visit(solidity::frontend::FunctionDefinition const&) (solc_ossfuzz+0xd6f07f)
    #30 0xe713cb in solidity::frontend::FunctionDefinition::accept(solidity::frontend::ASTConstVisitor&) const (solc_ossfuzz+0xe713cb)
    #31 0xd5b87a in solidity::frontend::TypeChecker::visit(solidity::frontend::ContractDefinition const&) (solc_ossfuzz+0xd5b87a)
    #32 0xe695fb in solidity::frontend::ContractDefinition::accept(solidity::frontend::ASTConstVisitor&) const (solc_ossfuzz+0xe695fb)
    #33 0xe66f1a in void solidity::frontend::ASTNode::listAccept<std::shared_ptr<solidity::frontend::ASTNode> >(std::vector<std::shared_ptr<solidity::frontend::ASTNode>, std::allocator<std::shared_ptr<solidity::frontend::ASTNode> > > const&, solidity::frontend::ASTConstVisitor&) (solc_ossfuzz+0xe66f1a)
    #34 0xe66985 in solidity::frontend::SourceUnit::accept(solidity::frontend::ASTConstVisitor&) const (solc_ossfuzz+0xe66985)
    #35 0xd59ec6 in solidity::frontend::TypeChecker::checkTypeRequirements(solidity::frontend::SourceUnit const&) (solc_ossfuzz+0xd59ec6)
    #36 0x8890f5 in solidity::frontend::CompilerStack::analyze() (solc_ossfuzz+0x8890f5)
    #37 0x892357 in solidity::frontend::CompilerStack::compile(solidity::frontend::CompilerStack::State) (solc_ossfuzz+0x892357)
    #38 0x56b6d5 in FuzzerUtil::testCompiler(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >&, bool, unsigned int, bool, bool) (solc_ossfuzz+0x56b6d5)
    #39 0x562ddf in LLVMFuzzerTestOneInput (solc_ossfuzz+0x562ddf)
    #40 0x543f49 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (solc_ossfuzz+0x543f49)
    #41 0x52ee49 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) (solc_ossfuzz+0x52ee49)
    #42 0x533d52 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (solc_ossfuzz+0x533d52)
    #43 0x52ebd2 in main (solc_ossfuzz+0x52ebd2)
    #44 0x7ffff7a620b2 in __libc_start_main /build/glibc-sMfBJT/glibc-2.31/csu/../csu/libc-start.c:308:16
    #45 0x48440d in _start (solc_ossfuzz+0x48440d)

NOTE: libFuzzer has rudimentary signal handlers.
      Combine libFuzzer with AddressSanitizer or similar for better crash reports.
SUMMARY: libFuzzer: deadly signal
```

zipped testcase to reproduce:
[id:000000,sig:06,src:012328,time:9919156,op:havoc,rep:2,trial:1.zip](https://github.com/ethereum/solidity/files/8592124/id.000000.sig.06.src.012328.time.9919156.op.havoc.rep.2.trial.1.zip)

**Comment by cameel** [comment]
Slightly simplified repro:
```solidity
contract C {
    function f() public {
        fixed8x80 x;
        assembly {
            let t := x
        }
    }
}
```
```bash
solc test.sol
```
```
Internal compiler error:
/solidity/libsolidity/analysis/TypeChecker.cpp(852): Throw in function solidity::frontend::TypeChecker::visit(const solidity::frontend::InlineAssembly&)::<lambda(const solidity::yul::Identifier&, solidity::yul::IdentifierContext, bool)>
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what: FixedPointType not implemented.
[solidity::util::tag_comment*] = FixedPointType not implemented.
```

This is a known problem though. Fixed point types like `fixed8x80` are already in the language but many of their aspects are simply not implemented (and likely won't be). The error reporting definitely should be better here (IMO it's not acceptable to just trigger an ICE for a missing feature) but that's all there is to it.

I think @bshastry had to deal with this too at some point and I think he's just making the fuzzer somehow ignore this particular problem.

**Comment by cameel** [comment]
Duplicate of #9936 and #9944.

**Comment by chriseth** [comment]
We could remove fixed points for 0.9.0.

**Comment by pietroborrello** [comment]
Thank you for looking into this. 
My bad! I assumed `fixed8x80` was a different "not implemented" type than `fixed` and thus, a different issue, or at least a different way to trigger the same underlying issue.
